### ISSUE TITLE
dunfell - Build error 

### ISSUE THREAD

**Issue by Oscar-Young** [munit-solutions/meta-azure-device-update#12]
I am following your [article ](https://githubmemory.com/repo/Azure/iot-hub-device-update/issues/79) and trying to build Yocto Image of Raspberry Pi 2

My bblayer.conf :

```
# POKY_BBLAYERS_CONF_VERSION is increased each time build/conf/bblayers.conf
# changes incompatibly
POKY_BBLAYERS_CONF_VERSION = "2"

BBPATH = "${TOPDIR}"
BBFILES ?= ""

BBLAYERS ?= " \
  /home/User01/poky/meta \
  /home/User01/poky/meta-poky \
  /home/User01/poky/meta-yocto-bsp \
  /home/User01/poky/meta-openembedded/meta-oe \
  /home/User01/poky/meta-openembedded/meta-networking \
  /home/User01/poky/meta-openembedded/meta-python \
  /home/User01/poky/meta-openembedded/meta-filesystems \
  /home/User01/poky/meta-raspberrypi \
  /home/User01/poky/meta-iotedge \
  /home/User01/poky/meta-rust \
  /home/User01/poky/meta-virtualization \
  /home/User01/poky/meta-swupdate \
  /home/User01/poky/meta-azure-device-update \
```

My local.conf :

```
#
# This file is your local configuration file and is where all local user settings
# are placed. The comments in this file give some guide to the options a new user
# to the system might want to change but pretty much any configuration option can
# be set in this file. More adventurous users can look at local.conf.extended
# which contains other examples of configuration which can be placed in this file
# but new users likely won't need any of them initially.
#
# Lines starting with the '#' character are commented out and in some cases the
# default values are provided as comments to show people example syntax. Enabling
# the option is a question of removing the # character and making any change to the
# variable as required.

#
# Machine Selection
#
# You need to select a specific machine to target the build with. There are a selection
# of emulated machines available which can boot and run in the QEMU emulator:
#
#MACHINE ?= "qemuarm"
#MACHINE ?= "qemuarm64"
#MACHINE ?= "qemumips"
#MACHINE ?= "qemumips64"
#MACHINE ?= "qemuppc"
#MACHINE ?= "qemux86"
#MACHINE ?= "qemux86-64"
#
# There are also the following hardware board target machines included for
# demonstration purposes:
#
#MACHINE ?= "beaglebone-yocto"
#MACHINE ?= "genericx86"
#MACHINE ?= "genericx86-64"
#MACHINE ?= "edgerouter"
#
# This sets the default machine to be qemux86-64 if no other machine is selected:
MACHINE ??= "raspberrypi2"

#
# Where to place downloads
#
# During a first build the system will download many different source code tarballs
# from various upstream projects. This can take a while, particularly if your network
# connection is slow. These are all stored in DL_DIR. When wiping and rebuilding you
# can preserve this directory to speed up this part of subsequent builds. This directory
# is safe to share between multiple builds on the same machine too.
#
# The default is a downloads directory under TOPDIR which is the build directory.
#
#DL_DIR ?= "${TOPDIR}/downloads"

#
# Where to place shared-state files
#
# BitBake has the capability to accelerate builds based on previously built output.
# This is done using "shared state" files which can be thought of as cache objects
# and this option determines where those files are placed.
#
# You can wipe out TMPDIR leaving this directory intact and the build would regenerate
# from these files if no changes were made to the configuration. If changes were made
# to the configuration, only shared state files where the state was still valid would
# be used (done using checksums).
#
# The default is a sstate-cache directory under TOPDIR.
#
#SSTATE_DIR ?= "${TOPDIR}/sstate-cache"
SDIMG_ROOTFS_TYPE = "ext4"
RPI_USE_U_BOOT = "1"
ENABLE_UART = "1"
PREFERRED_PROVIDER_virtual/bootloader = "u-boot"

IMAGE_FSTYPES += "wic wic.bmap"
PREFERRED_PROVIDER_u-boot-fw-utils = "libubootenv"
ADUC_PRIVATE_KEY = "/home/User01/priv.pem"
ADUC_PRIVATE_KEY_PASSWORD = "/home/User01/priv.pass"

#
# Where to place the build output
#
# This option specifies where the bulk of the building work should be done and
# where BitBake should place its temporary files and output. Keep in mind that
# this includes the extraction and compilation of many applications and the toolchain
# which can use Gigabytes of hard disk space.
#
# The default is a tmp directory under TOPDIR.
#
#TMPDIR = "${TOPDIR}/tmp"

#
# Default policy config
#
# The distribution setting controls which policy settings are used as defaults.
# The default value is fine for general Yocto project use, at least initially.
# Ultimately when creating custom policy, people will likely end up subclassing
# these defaults.
#
DISTRO ?= "poky"
# As an example of a subclass there is a "bleeding" edge policy configuration
# where many versions are set to the absolute latest code from the upstream
# source control systems. This is just mentioned here as an example, its not
# useful to most new users.
# DISTRO ?= "poky-bleeding"

#
# Package Management configuration
#
# This variable lists which packaging formats to enable. Multiple package backends
# can be enabled at once and the first item listed in the variable will be used
# to generate the root filesystems.
# Options are:
#  - 'package_deb' for debian style deb files
#  - 'package_ipk' for ipk files are used by opkg (a debian style embedded package manager)
#  - 'package_rpm' for rpm style packages
# E.g.: PACKAGE_CLASSES ?= "package_rpm package_deb package_ipk"
# We default to rpm:
PACKAGE_CLASSES ?= "package_rpm"

#
# SDK target architecture
#
# This variable specifies the architecture to build SDK items for and means
# you can build the SDK packages for architectures other than the machine you are
# running the build on (i.e. building i686 packages on an x86_64 host).
# Supported values are i686 and x86_64
#SDKMACHINE ?= "i686"

#
# Extra image configuration defaults
#
# The EXTRA_IMAGE_FEATURES variable allows extra packages to be added to the generated
# images. Some of these options are added to certain image types automatically. The
# variable can contain the following options:
#  "dbg-pkgs"       - add -dbg packages for all installed packages
#                     (adds symbol information for debugging/profiling)
#  "src-pkgs"       - add -src packages for all installed packages
#                     (adds source code for debugging)
#  "dev-pkgs"       - add -dev packages for all installed packages
#                     (useful if you want to develop against libs in the image)
#  "ptest-pkgs"     - add -ptest packages for all ptest-enabled packages
#                     (useful if you want to run the package test suites)
#  "tools-sdk"      - add development tools (gcc, make, pkgconfig etc.)
#  "tools-debug"    - add debugging tools (gdb, strace)
#  "eclipse-debug"  - add Eclipse remote debugging support
#  "tools-profile"  - add profiling tools (oprofile, lttng, valgrind)
#  "tools-testapps" - add useful testing tools (ts_print, aplay, arecord etc.)
#  "debug-tweaks"   - make an image suitable for development
#                     e.g. ssh root access has a blank password
# There are other application targets that can be used here too, see
# meta/classes/image.bbclass and meta/classes/core-image.bbclass for more details.
# We default to enabling the debugging tweaks.
EXTRA_IMAGE_FEATURES ?= "debug-tweaks ssh-server-dropbear tools-debug tools-sdk"

#
# Additional image features
#
# The following is a list of additional classes to use when building images which
# enable extra features. Some available options which can be included in this variable
# are:
#   - 'buildstats' collect build statistics
#   - 'image-mklibs' to reduce shared library files size for an image
#   - 'image-prelink' in order to prelink the filesystem image
# NOTE: if listing mklibs & prelink both, then make sure mklibs is before prelink
# NOTE: mklibs also needs to be explicitly enabled for a given image, see local.conf.extended
USER_CLASSES ?= "buildstats image-mklibs image-prelink"

#
# Runtime testing of images
#
# The build system can test booting virtual machine images under qemu (an emulator)
# after any root filesystems are created and run tests against those images. It can also
# run tests against any SDK that are built. To enable this uncomment these lines.
# See classes/test{image,sdk}.bbclass for further details.
#IMAGE_CLASSES += "testimage testsdk"
#TESTIMAGE_AUTO_qemuall = "1"

#
# Interactive shell configuration
#
# Under certain circumstances the system may need input from you and to do this it
# can launch an interactive shell. It needs to do this since the build is
# multithreaded and needs to be able to handle the case where more than one parallel
# process may require the user's attention. The default is iterate over the available
# terminal types to find one that works.
#
# Examples of the occasions this may happen are when resolving patches which cannot
# be applied, to use the devshell or the kernel menuconfig
#
# Supported values are auto, gnome, xfce, rxvt, screen, konsole (KDE 3.x only), none
# Note: currently, Konsole support only works for KDE 3.x due to the way
# newer Konsole versions behave
#OE_TERMINAL = "auto"
# By default disable interactive patch resolution (tasks will just fail instead):
PATCHRESOLVE = "noop"

#
# Disk Space Monitoring during the build
#
# Monitor the disk space during the build. If there is less that 1GB of space or less
# than 100K inodes in any key build location (TMPDIR, DL_DIR, SSTATE_DIR), gracefully
# shutdown the build. If there is less that 100MB or 1K inodes, perform a hard abort
# of the build. The reason for this is that running completely out of space can corrupt
# files and damages the build in ways which may not be easily recoverable.
# It's necesary to monitor /tmp, if there is no space left the build will fail
# with very exotic errors.
BB_DISKMON_DIRS ??= "\
    STOPTASKS,${TMPDIR},1G,100K \
    STOPTASKS,${DL_DIR},1G,100K \
    STOPTASKS,${SSTATE_DIR},1G,100K \
    STOPTASKS,/tmp,100M,100K \
    ABORT,${TMPDIR},100M,1K \
    ABORT,${DL_DIR},100M,1K \
    ABORT,${SSTATE_DIR},100M,1K \
    ABORT,/tmp,10M,1K"

#
# Shared-state files from other locations
#
# As mentioned above, shared state files are prebuilt cache data objects which can
# used to accelerate build time. This variable can be used to configure the system
# to search other mirror locations for these objects before it builds the data itself.
#
# This can be a filesystem directory, or a remote url such as http or ftp. These
# would contain the sstate-cache results from previous builds (possibly from other
# machines). This variable works like fetcher MIRRORS/PREMIRRORS and points to the
# cache locations to check for the shared objects.
# NOTE: if the mirror uses the same structure as SSTATE_DIR, you need to add PATH
# at the end as shown in the examples below. This will be substituted with the
# correct path within the directory structure.
#SSTATE_MIRRORS ?= "\
#file://.* http://someserver.tld/share/sstate/PATH;downloadfilename=PATH \n \
#file://.* file:///some/local/dir/sstate/PATH"

#
# Yocto Project SState Mirror
#
# The Yocto Project has prebuilt artefacts available for its releases, you can enable
# use of these by uncommenting the following line. This will mean the build uses
# the network to check for artefacts at the start of builds, which does slow it down
# equally, it will also speed up the builds by not having to build things if they are
# present in the cache. It assumes you can download something faster than you can build it
# which will depend on your network.
#
#SSTATE_MIRRORS ?= "file://.* http://sstate.yoctoproject.org/2.5/PATH;downloadfilename=PATH"

#
# Qemu configuration
#
# By default native qemu will build with a builtin VNC server where graphical output can be
# seen. The line below enables the SDL UI frontend too.
PACKAGECONFIG_append_pn-qemu-system-native = " sdl"
# By default libsdl2-native will be built, if you want to use your host's libSDL instead of
# the minimal libsdl built by libsdl2-native then uncomment the ASSUME_PROVIDED line below.
#ASSUME_PROVIDED += "libsdl2-native"

# You can also enable the Gtk UI frontend, which takes somewhat longer to build, but adds
# a handy set of menus for controlling the emulator.
#PACKAGECONFIG_append_pn-qemu-system-native = " gtk+"

#
# Hash Equivalence
#
# Enable support for automatically running a local hash equivalence server and
# instruct bitbake to use a hash equivalence aware signature generator. Hash
# equivalence improves reuse of sstate by detecting when a given sstate
# artifact can be reused as equivalent, even if the current task hash doesn't
# match the one that generated the artifact.
#
# A shared hash equivalent server can be set with "<HOSTNAME>:<PORT>" format
#
#BB_HASHSERVE = "auto"
#BB_SIGNATURE_HANDLER = "OEEquivHash"

# CONF_VERSION is increased each time build/conf/ changes incompatibly and is used to
# track the version of this file when it was generated. This can safely be ignored if
# this doesn't mean anything to you.
CONF_VERSION = "1"
IMAGE_INSTALL_append += " \
  iotedge-daemon \
  iotedge-cli \
  docker \
  ca-certificates \
  connman \
  connman-client \
  azure-device-update \
  adu-agent-service \
  "

DISTRO_FEATURES_append += " systemd wifi virtualization"
VIRTUAL-RUNTIME_init_manager = "systemd"
DISTRO_FEATURES_BACKFILL_CONSIDERED = "sysvinit"
VIRTUAL-RUNTIME_initscripts = ""
```

When `bitbake core-image-base` t I get some error messages :

```
Loading cache: 100% |####################################################################################| Time: 0:00:00Loaded 3473 entries from dependency cache.
WARNING: /home/User01/poky/meta-azure-device-update/recipes-extended/images/adu-update-image.bb: Unable to get checksum for adu-update-image SRC_URI entry sw-description: file could not be found
WARNING: /home/User01/poky/meta-azure-device-update/recipes-extended/images/adu-update-image.bb: Unable to get checksum for adu-update-image SRC_URI entry manifest.json.template: file could not be found
Parsing recipes: 100% |##################################################################################| Time: 0:00:01Parsing of 2333 .bb files complete (2308 cached, 25 parsed). 3475 targets, 137 skipped, 0 masked, 0 errors.
Removing 1 recipes from the cortexa7t2hf-neon-vfpv4 sysroot: 100% |######################################| Time: 0:00:01NOTE: Resolving any missing task queue dependencies

Build Configuration:
BB_VERSION           = "1.46.0"
BUILD_SYS            = "x86_64-linux"
NATIVELSBSTRING      = "universal"
TARGET_SYS           = "arm-poky-linux-gnueabi"
MACHINE              = "raspberrypi2"
DISTRO               = "poky"
DISTRO_VERSION       = "3.1.12"
TUNE_FEATURES        = "arm vfp cortexa7 neon vfpv4 thumb callconvention-hard"
TARGET_FPU           = "hard"
meta
meta-poky
meta-yocto-bsp       = "dunfell:cf5a00721f721d5077c73d1f4e812e5c79833fba"
meta-oe
meta-networking
meta-python
meta-filesystems     = "dunfell:69f94af4d91215e7d4e225bab54bf3bcfee42f1c"
meta-raspberrypi     = "dunfell:934064a01903b2ba9a82be93b3f0efdb4543a0e8"
meta-iotedge         = "dunfell:a10eaf625500f02049b76ccfc02378d845cbb442"
meta-rust            = "HEAD:7ff669d8cedd83a2d3efb73073a63b0a7efffddc"
meta-virtualization  = "dunfell:180241e0eee718fc52c7b6b25dbd1d845a8047c4"
meta-swupdate        = "dunfell:fe937d7deb593040bcb785eb3080bb166c34b8b2"
meta-azure-device-update = "dunfell:539d7031babff21ecc4ab052b318c51ddb432551"

Initialising tasks: 100% |###############################################################################| Time: 0:00:07Sstate summary: Wanted 49 Found 0 Missed 49 Current 2173 (0% match, 97% complete)
NOTE: Executing Tasks
WARNING: azure-iot-sdk-c-1.0+gitAUTOINC+6b6e244967-r0 do_fetch: Failed to fetch URL gitsm://github.com/Azure/azure-iot-sdk-c.git;branch=master, attempting MIRRORS if available
ERROR: azure-iot-sdk-c-1.0+gitAUTOINC+6b6e244967-r0 do_fetch: Fetcher failure: Unable to find revision 6b6e2449672d229d4a0fc9503d8fd3b2b2dde5f1 in branch master even from upstream
ERROR: azure-iot-sdk-c-1.0+gitAUTOINC+6b6e244967-r0 do_fetch: Bitbake Fetcher Error: FetchError('Unable to fetch URL from any source.', 'gitsm://github.com/Azure/azure-iot-sdk-c.git;branch=master')
ERROR: Logfile of failure stored in: /home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+gitAUTOINC+6b6e244967-r0/temp/log.do_fetch.3397971
ERROR: Task (/home/User01/poky/meta-azure-device-update/recipes-azure-iot/azure-iot-sdk-c/azure-iot-sdk-c_git.bb:do_fetch) failed with exit code '1'
Waiting for 1 running tasks to finish:
```

Then i try to replace azure-iot.sdk-c_git.bb SRCRED and SRC_URI and got another error.

```
Initialising tasks: 100% |###############################################################| Time: 0:00:07
Sstate summary: Wanted 51 Found 0 Missed 51 Current 2171 (0% match, 97% complete)
NOTE: Executing Tasks
ERROR: azure-iot-sdk-c-1.0+git-r0 do_configure: Execution of '/home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+git-r0/temp/run.do_configure.3396797' failed with exit code 1
ERROR: Logfile of failure stored in: /home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+git-r0/temp/log.do_configure.3396797
Log data follows:
| DEBUG: Executing shell function do_configure
| -- The C compiler identification is GNU 9.3.0
| -- The CXX compiler identification is GNU 9.3.0
| -- Check for working C compiler: /home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+git-r0/recipe-sysroot-native/usr/bin/arm-poky-linux-gnueabi/arm-poky-linux-gnueabi-gcc
| -- Check for working C compiler: /home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+git-r0/recipe-sysroot-native/usr/bin/arm-poky-linux-gnueabi/arm-poky-linux-gnueabi-gcc -- works
| -- Detecting C compiler ABI info
| -- Detecting C compiler ABI info - done
| -- Detecting C compile features
| -- Detecting C compile features - done
| -- Check for working CXX compiler: /home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+git-r0/recipe-sysroot-native/usr/bin/arm-poky-linux-gnueabi/arm-poky-linux-gnueabi-g++
| -- Check for working CXX compiler: /home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+git-r0/recipe-sysroot-native/usr/bin/arm-poky-linux-gnueabi/arm-poky-linux-gnueabi-g++ -- works
| -- Detecting CXX compiler ABI info
| -- Detecting CXX compiler ABI info - done
| -- Detecting CXX compile features
| -- Detecting CXX compile features - done
| -- IoT Client SDK Version = 1.6.0
| CMake Error at CMakeLists.txt:259 (add_subdirectory):
|   The source directory
|
|     /home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+git-r0/git/deps/umock-c
|
|   does not contain a CMakeLists.txt file.
|
|
| CMake Error at dependencies.cmake:22 (add_subdirectory):
|   The source directory
|
|     /home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+git-r0/git/c-utility
|
|   does not contain a CMakeLists.txt file.
| Call Stack (most recent call first):
|   CMakeLists.txt:274 (include)
|
|
| CMake Error at dependencies.cmake:29 (add_subdirectory):
|   The source directory
|
|     /home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+git-r0/git/umqtt
|
|   does not contain a CMakeLists.txt file.
| Call Stack (most recent call first):
|   CMakeLists.txt:274 (include)
|
|
| CMake Error at dependencies.cmake:33 (add_subdirectory):
|   The source directory
|
|     /home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+git-r0/git/deps/uhttp
|
|   does not contain a CMakeLists.txt file.
| Call Stack (most recent call first):
|   CMakeLists.txt:274 (include)
|
|
| CMake Error at CMakeLists.txt:285 (set_platform_files):
|   Unknown CMake command "set_platform_files".
|
|
| -- Configuring incomplete, errors occurred!
| See also "/home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+git-r0/build/CMakeFiles/CMakeOutput.log".
| WARNING: exit code 1 from a shell command.
| ERROR: Execution of '/home/User01/poky/build/tmp/work/cortexa7t2hf-neon-vfpv4-poky-linux-gnueabi/azure-iot-sdk-c/1.0+git-r0/temp/run.do_configure.3396797' failed with exit code 1
ERROR: Task (/home/User01/poky/meta-azure-device-update/recipes-azure-iot/azure-iot-sdk-c/azure-iot-sdk-c_git.bb:do_configure) failed with exit code '1'
```

I am not good at C and Yocto, So I am not clear what this error means. Is this what dependency package I am missing ?

**Comment by SimonSimCity** [comment]
The first error you reported should be fixed by a PR I merged today (#9). Have you solved it in this way? If not, please check if the layer now works as expected.

**Comment by Oscar-Young** [comment]
Hi @SimonSimCity ,

Thanks, compiled successfully !
